### ISSUE TITLE
Add "beginner's workflow section"

### ISSUE THREAD

**Issue by achillemyette1** [mila-iqia/mila-docs#11]
(This issue was originally raised by user **vict0rsch** in the previous repository "AI-HPC-Docs" - on Oct 21, 2020)

Original Link: https://github.com/mila-iqia/AI-HPC-Docs/issues/73

@vict0rsch
vict0rsch commented on Oct 21, 2020
Say someone arrives at Mila with a basic understanding of python environments and has only worked on their laptop, how should they get started? something along the lines of this document I made

@vict0rsch
Author
vict0rsch commented on Oct 21, 2020
Connecting to Mila’s Cluster
Before starting: get your Mila cluster account ready. Should look something like surname[:6]firstname[0] but this may vary, it’s not a necessary condition.

ssh
Install an ssh client on your computer. Then:

#generic login will send you to one of the 4 login nodes to spread the load
ssh user@login.server.mila.quebec -p 2222
#loginX, X in [1, 2, 3, 4]
ssh user@login-X.login.server.mila.quebec -p 2222
You can also use a config file > https://mila-umontreal.slack.com/archives/C2TL9FRBP/p1581462695079500 to just do ssh mila

You can use these credentials to

Cluster Usage
Login Nodes
Affter ssh you’ll land on a so-called login node (1, 2, 3 or 4). Don’t ever run anything compute-heavy on it. You could bring the node down, preventing everyone else from accessing the cluster.

This means no python script or anything like that, but also no zip on large folders and so on. Nothing that’s computationally heavy. Rule of thumbs: execution can’t take more than a few seconds.

In any case, read the Mila cluster documentation => docs.mila.quebec

Compute Nodes
Your computations should be done on a compute node, which you request through SLURM. This software manages requests and allocates ressources to the cluster's hundreds of users.

run kinit to make sure your "ticket" is up to date (docs link)
run sbatch job.sh to put your job in the queue (docs link)
run srun <options> to get an interactive session and play around with the compute node with a terminal in it (docs link)
go to https://jupyterhub.server.mila.quebec to get notebooks (docs link)
savail will show you currenlty available GPUs.

Partitions
Jobs are allocated according to their priority in the queue of requested jobs by all users. You can request one of 3 types of allocations:

unkillable is the most stable: you only get 1 job with such a partition. Jobs requesting -p unkillable will not be preempted (=stopped to give the GPU to another job with higher priority). You have this node for up to 24h. You cannot request more than 4 cpus 48GB of RAM and 1 GPU
main is a high priority job: it may be preempted but it's very unlikely: you get 2 jobs with such a partition. You have this node for up to 48h and the total cpus requested by your main jobs cannot exceed 12.
long is a low priority job: it may be preempted so you should have regular checkpoints. There are no limits on the resources you can request, but bear in mind that the more resources you request, the more likely that they will be needed by a higher priority job and yours will be killed to release them.
If you do not specify -p (or, equivalently —partition) then the default is long.

Python
You should use anaconda, which is available on the cluster through the module tool (docs link)

$ module load anaconda/3
$ source $CONDA_ACTIVATE
(base) $ conda create --name 'myenv' python=3.7.4
(base) $ conda activate myenv
(myenv) $ pip install -r requirements.txt
If you have Cuda issues, remember you can load the appropriate Cuda and CuDnn versions from module

Examples
srun --gres=gpu:titanxp:1 --cpus-per-task=4 --mem=16GB --pty bash #starts an interactive job with a bash shell
job.sh:

#!/bin/bash
#SBATCH --cpus-per-task=4  # request cpus
#SBATCH --gres=gpu:1:titanxp  # request 1 titanxp
#SBATCH --mem=16GB  # RAM memory, NOT gpu memory
#SBATCH -o /network/tmp1/<user>/slurm_outputs/slurm-%j.out  # will write the prints etc. in this file where `%j` will be the job's id (like 328892)
#SBATCH -p main # partition

cd $SLURM_TMPDIR && cp path/to/data.zip . && unzip data.zip # best pratcice, may be too long depending on your data's size

module load anaconda/3 >/dev/null 2>&1
. "$CONDA_ACTIVATE"
conda activate myenv

echo "Starting job"

cd path/to/code

python train.py --data_path=$SLURM_TMPDIR

echo 'done'
$SLURM_TMPDIR is the path to a large, I/O efficient temporary folder attached to the compute node for the current job only so you should read and write from there but remember to move anything you want to keep to a permanent storage space like /network/tmp1 or /miniscratch.

Storing data
Imporant read: https://docs.mila.quebec/mila-cluster/index.html#milacluster-storage

TL;DR:

Only put code and packages in $HOME = ~ = /newtwork/home/<user>
Store large chunks of data (processed datasets, checkpoints etc.) in tmp1: /network/tmp1/<user> or /miniscratch/<user>
Do your computations on the temporary directory that you get when allocated a compute node: $SLURM_TMPDIR
@vict0rsch
Author
vict0rsch commented on Oct 21, 2020
I give that to my interns. We should have one for Mila

**Comment by vict0rsch** [comment]
Formatted:

**Comment by vict0rsch** [comment]
# Connecting to Mila’s Cluster

Before starting: get your Mila cluster account ready. Should look something like `surname[:6]firstname[0]` but this may vary, it’s not a necessary condition.

## ssh

Install an `ssh` client on your computer. Then:

```bash
#generic login will send you to one of the 4 login nodes to spread the load
ssh user@login.server.mila.quebec -p 2222
#loginX, X in [1, 2, 3, 4]
ssh user@login-X.login.server.mila.quebec -p 2222
```

You can also use a config file > https://mila-umontreal.slack.com/archives/C2TL9FRBP/p1581462695079500 to just do `ssh mila`

You can use these credentials to

# Cluster Usage

## Login Nodes

Affter `ssh` you’ll land on a so-called login node (`1`, `2`, `3` or `4`). Don’t **ever** run anything compute-heavy on it. You could bring the node down, preventing everyone else from accessing the cluster.  

This means no `python` script or anything like that, but also no `zip` on large folders and so on. Nothing that’s computationally heavy. Rule of thumbs: execution can’t take more than a few seconds.

In any case, **read the Mila cluster documentation => [docs.mila.quebec](docs.mila.quebec)**

## Compute Nodes

Your computations should be done on a compute node, which you request through SLURM. This software manages requests and allocates ressources to the cluster's hundreds of users.

1. run `kinit` to make sure your "ticket" is up to date ([docs link](https://docs.mila.quebec/mila-cluster/index.html#milacluster-storage))
2. run `sbatch job.sh` to put your job in the queue ([docs link](https://docs.mila.quebec/mila-cluster/index.html#script-example))
3. run `srun <options>` to get an interactive session and play around with the compute node with a terminal in it ([docs link](https://docs.mila.quebec/mila-cluster/index.html#launching-jobs))
4. go to https://jupyterhub.server.mila.quebec to get notebooks ([docs link](https://jupyterhub.server.mila.quebec))

`savail` will show you currenlty available GPUs.

### Partitions

Jobs are allocated according to their priority in the queue of requested jobs by all users. You can request one of 3 types of allocations:

1. `unkillable` is the most stable: you only get 1 job with such a partition. Jobs requesting `-p unkillable` will not be preempted (=stopped to give the GPU to another job with higher priority). You have this node for up to 24h. You cannot request more than 4 cpus 48GB of RAM and 1 GPU
2. `main` is a high priority job: it may be preempted but it's very unlikely: you get 2 jobs with such a partition. You have this node for up to 48h and the total cpus requested by your `main`  jobs cannot exceed 12.
3. `long` is a low priority job: it may be preempted so you should have regular checkpoints. There are no limits on the resources you can request, but bear in mind that the more resources you request, the more likely that they will be needed by a higher priority job and yours will be killed to release them.

If you do not specify `-p` (or, equivalently `—partition`) then the default is `long`.

### Python

You should use `anaconda`, which is available on the cluster through the `module` tool ([docs link](https://docs.mila.quebec/software/index.html#modules))

1. `$ module load anaconda/3`
2. `$ source $CONDA_ACTIVATE`
3. `(base) $ conda create --name 'myenv' python=3.7.4 `
4. `(base) $ conda activate myenv `
5. `(myenv) $ pip install -r requirements.txt`

If you have Cuda issues, remember you can load the appropriate Cuda and CuDnn versions from `module`

###### Examples

```bash
srun --gres=gpu:titanxp:1 --cpus-per-task=4 --mem=16GB --pty bash #starts an interactive job with a bash shell
```

`job.sh`:

```bash
#!/bin/bash
#SBATCH --cpus-per-task=4  # request cpus
#SBATCH --gres=gpu:1:titanxp  # request 1 titanxp
#SBATCH --mem=16GB  # RAM memory, NOT gpu memory
#SBATCH -o /network/tmp1/<user>/slurm_outputs/slurm-%j.out  # will write the prints etc. in this file where `%j` will be the job's id (like 328892)
#SBATCH -p main # partition

cd $SLURM_TMPDIR && cp path/to/data.zip . && unzip data.zip # best pratcice, may be too long depending on your data's size

module load anaconda/3 >/dev/null 2>&1
. "$CONDA_ACTIVATE"
conda activate myenv

echo "Starting job"

cd path/to/code

python train.py --data_path=$SLURM_TMPDIR

echo 'done'

```

 `$SLURM_TMPDIR` is the path to a large, I/O efficient **temporary** folder attached to the compute node **for the current job only** so you should read and write from there but remember to **move anything you want to keep** to a permanent storage space like `/network/tmp1` or `/miniscratch`.

### Storing data

**Imporant** read: https://docs.mila.quebec/mila-cluster/index.html#milacluster-storage

TL;DR: 

1. Only put code and packages in `$HOME` = `~` = `/newtwork/home/<user>`
2. Store large chunks of data (processed datasets, checkpoints etc.) in `tmp1`: `/network/tmp1/<user>` or `/miniscratch/<user>`
3. Do your computations on the temporary directory that you get when allocated a compute node: `$SLURM_TMPDIR`
